src/doc/performance-tuning.md

```diff
@@ -1,91 +1,185 @@
+performance", "Energy Efficiency" and "Custom".  Select "Max
+performance" for latency sensitive Snabb use.
+
+### Disable hyperthreads
+
+Hyperthreads are a way of maximizing resource utilization on a CPU core,
+driven by the observation that a CPU is often waiting on memory or some
+external event, and might as well be doing something else while it's
+waiting.  In such a situation, it can be advantageous to run a second
+thread on that CPU.  However for Snabb that's exactly what we don't
+want.  We do not want another thread competing for compute and cache
+resources on our CPU and increasing our latency.  For best results and
+lowest latency, disable hyperthreading via the BIOS settings.
+
+### Consider disabling Turbo Boost
+
+Intel Turbo Boost Technology allows processor cores to run faster than
+the rated operating frequency if they're operating below power,
+current, and temperature specification limits.  The conservative thing
+to do is to turn off Turbo Boost when doing performance tests of Snabb
+data planes, because that way you reduce variance due to CPU clock
+speed changing.
+
+## System settings
+
+### Adjust CPU frequency governor
+
+To avoid power-saving heuristics causing decreased throughput and higher
+latency, set the CPU frequency governor to `performance`:
+
+```bash
+for CPUFREQ in /sys/devices/system/cpu/cpu*/cpufreq/scaling_governor; do
+   [ -f $CPUFREQ ] || continue;
+   echo -n performance > $CPUFREQ;
+done
 ```
 
-The default of 512 seems too small, based on load test at IMIX line rate tests against lwaftr, 1024 or 2048 gave equally good results. Num_descriptors controls the Receive Descriptor Length on the Intel 82599 Controller, which determines the number of packets allocated to the circular buffer. This value must be a power of two. Larger ingress buffer can reduce packet loss while Snabb is busy handling other packets, but it will also increase latency for packets waiting in the queue to be picked up by Snabb.
+### Reserve CPUs for data plane usage
 
-### Enable engine.busywait
-Defined in src/core/app.lua and enabled before calling engine.main() via
+When running a Snabb data plane, we don't want interference from the
+Linux kernel.  In normal operation, a Snabb data plane won't even make
+any system calls at all.  You can prevent the Linux kernel from
+pre-empting your Snabb application to schedule other processes on its
+CPU by reserving CPUs via the `isolcpus` kernel boot setting.
+
+To isolate CPUs, boot your Linux kernel with the `isolcpus` parameter.
+Under NixOS, edit `/etc/nixos/configuration.nix` to add this parameter:
 
 ```
-engine.busywait = true
-engine.main(...)
+boot.kernelParams = [ "isolcpus=1-5,7-11" ];
 ```
-If true then the engine will poll for new data in a tight loop (100% CPU) instead of sleeping according to the Hz setting. This will reduce overall packet latency and increase throughput at the cost of utilizing the CPU hosting Snabb at 100%. 
 
-### Monitor ifInDiscards
```

Comment:
Yes.  IMO people should use the ingress drop monitor or the ingress drop counter, mentioned later.